test/e2e/azure_json.go

```diff
@@ -0,0 +1,113 @@
+//go:build e2e
+// +build e2e
+
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/cluster-api-provider-azure/controllers"
+	"sigs.k8s.io/cluster-api/test/framework"
+)
+
+const (
+	getAzureJSONInterval = 5 * time.Second
+	getAzureJSONTimeout  = 2 * time.Minute
+	guidRegex            = `^[{]?[0-9a-fA-F]{8}-([0-9a-fA-F]{4}-){3}[0-9a-fA-F]{12}[}]?$`
+)
+
+// AzureJSONSecretSpecInput is the input for AzureJSONSecretSpec.
+type AzureJSONSecretSpecInput struct {
+	BootstrapClusterProxy framework.ClusterProxy
+	Namespace             *corev1.Namespace
+	ClusterName           string
+}
+
+// AzureJSONSecretSpec implements a test that verifies the contents of the /etc/kubernetes/azure.json file.
+func AzureJSONSecretSpec(ctx context.Context, inputGetter func() AzureJSONSecretSpecInput) {
```

Comment:
I was thinking we should get rid of this test completely in favor of the LB test. 

If you still want to do the extra validation step of checking the content of the AzureJson file, consider making this function more specific to what we are validating which is identity. With the current implementation, it assumes user-assigned identity even though the func name and comment make it seem like it's a generic test for azure.json content. It also checks for hardcoded values like `LoadBalancerSku` being "Regular" which might not apply to every cluster if we ever want to reuse this test spec. It would also be good to make sure that ClientID / ClientSecret are empty when using managed identity, and vice versa when using a service principal.

Reply:
> I was thinking we should get rid of this test completely in favor of the LB test.

That's fine with me. I just need to validate that the LB Spec + user-assigned ID test covers this and works.

I had meant this test to do a more general validation of `azure.json` based on the cluster configuration, but that's not specific to this case. Also I'm worried that involving SSH could introduce more flakiness into the tests overall, and we don't need that.

Reply:
Looking through test again, I'm more convinced we should get rid of this test. If we want to test the content of the file, it should be covered via unit test (https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/controllers/helpers_test.go#L82). The goal of the e2e test is to make sure the cluster is functional. In this case a cluster using user-assigned identities should provide valid credentials for cloud provider to make Azure API calls. That is already validated by the LB spec.